src/hnsw/build.c

```diff
@@ -482,25 +492,50 @@ static void BuildIndex(
         assert(error == NULL);
     }
 
-    char *result_buf = NULL;
-    usearch_save(buildstate->usearch_index, NULL, &result_buf, &error);
-    assert(error == NULL && result_buf != NULL);
-
-    size_t num_added_vectors = usearch_size(buildstate->usearch_index, &error);
+    metadata = usearch_metadata(buildstate->usearch_index, &error);
     assert(error == NULL);
 
+    if(buildstate->index_file_path == NULL) {
+        // size of static name + max digits of uint32 (Oid) + 1 for nullbyte
+        index_file_path = palloc0(sizeof("/tmp/ldb-index-.bin") + 10 + 1);
+        sprintf(index_file_path, "/tmp/ldb-index-%d.bin", index->rd_rel->relfilenode);
+        usearch_save(buildstate->usearch_index, index_file_path, NULL, &error);
+        assert(error == NULL);
+    } else {
+        index_file_path = buildstate->index_file_path;
+    }
+
+    num_added_vectors = usearch_size(buildstate->usearch_index, &error);
+    assert(error == NULL);
     elog(INFO, "done saving %ld vectors", num_added_vectors);
 
-    //****************************** saving to WAL BEGIN ******************************//
-    UpdateProgress(PROGRESS_CREATEIDX_PHASE, PROGRESS_HNSW_PHASE_LOAD);
-    StoreExternalIndex(index, buildstate->usearch_index, forkNum, result_buf, &opts, num_added_vectors);
+    //****************************** mmap index to memory BEGIN ******************************//
+    index_file_fd = open(index_file_path, O_RDONLY);
+    assert(index_file_fd > 0);
 
-    //****************************** saving to WAL END ******************************//
+    fstat(index_file_fd, &index_file_stat);
+    result_buf = mmap(NULL, index_file_stat.st_size, PROT_READ, MAP_PRIVATE, index_file_fd, 0);
+    assert(result_buf != MAP_FAILED);
 
     usearch_free(buildstate->usearch_index, &error);
-    free(result_buf);
     assert(error == NULL);
     buildstate->usearch_index = NULL;
+    //****************************** mmap index to memory END ******************************//
+
+    //****************************** saving to WAL BEGIN ******************************//
+    UpdateProgress(PROGRESS_CREATEIDX_PHASE, PROGRESS_HNSW_PHASE_LOAD);
+    StoreExternalIndex(index, &metadata, forkNum, result_buf, &opts, num_added_vectors);
+
+    //****************************** saving to WAL END ******************************//
+
+    munmap(result_buf, index_file_stat.st_size);
+    close(index_file_fd);
+
+    if(buildstate->index_file_path == NULL) {
+        // remove index file if it was not externally provided
+        unlink(index_file_path);
```

Comment:
can you store the tmp index_file_path in another var called `tmp_inde_file_path`, and make this branch conditional on `sprintf`ed `tmp_inde_file_path` existing?
The way it stands, it is confusing and I am afraid some refactoring may end us up deleting users' externally managed files